internal/metadata/metadata.go

```diff
@@ -0,0 +1,80 @@
+// Copyright 2016-2023, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metadata
+
+import (
+	"encoding/json"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
+)
+
+// The underlying value of a metadata blob.
+type Data struct {
+	m map[string]json.RawMessage
+}
+
+func New(data []byte) (*Data, error) {
+	m := map[string]json.RawMessage{}
+	if data != nil {
+		err := json.Unmarshal(data, &m)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return &Data{m}, nil
+}
+
+func (d *Data) Marshal() []byte {
+	if d == nil {
+		d = &Data{m: make(map[string]json.RawMessage)}
+	}
+	bytes, err := json.MarshalIndent(d.m, "", "    ")
+	contract.AssertNoErrorf(err, "internal: failed to marshal metadata")
```

Comment:
My only nit on this PR is the use of `contract`. 

I think we had been moving away from panics in non-test code, in light of how panics hid errors lurking in docsgen, which used `recover()`.

Reply:
I'll add a note explaining why the `contract` assertion is there, but I think it should stay. In my mind, I separate error handling (something has gone wrong, let's deal with it) from assertions (this should never happen, and if it does I don't want to deal with it. Let the compiler know). The second type shouldn't be returned as errors, but I agree that there needs to be a better explanation in the code.

Here, the explanation is
```
`d.m` is a `map[string]json.RawMessage`. `json.MarshalIndent` errors only when it is 
asked to serialize an unmarshalable type (complex, function or channel) or a cyclic data
structure. Because `string` and `json.RawMessage` are trivially marshallable and cannot
contain cycles, all values of `d.m` can be marshaled without error. 
```